zcash_client_backend/src/data_api.rs

```diff
@@ -0,0 +1,264 @@
+        self.block_height_extrema().map(|heights| {
+            heights.map(|(min_height, max_height)| {
+                let target_height = max_height + 1;
+
+                // Select an anchor ANCHOR_OFFSET back from the target block,
+                // unless that would be before the earliest block we have.
+                let anchor_height = BlockHeight::from(cmp::max(
+                    u32::from(target_height).saturating_sub(ANCHOR_OFFSET),
+                    u32::from(min_height),
+                ));
+
+                (target_height, anchor_height)
+            })
+        })
+    }
+
+    fn get_block_hash(&self, block_height: BlockHeight) -> Result<Option<BlockHash>, Self::Error>;
+
+    fn get_max_height_hash(&self) -> Result<Option<(BlockHeight, BlockHash)>, Self::Error> {
+        self.block_height_extrema()
+            .and_then(|extrema_opt| {
+                extrema_opt
+                    .map(|(_, max_height)| {
+                        self.get_block_hash(max_height)
+                            .map(|hash_opt| hash_opt.map(move |hash| (max_height, hash)))
+                    })
+                    .transpose()
+            })
+            .map(|oo| oo.flatten())
+    }
+
+    fn get_tx_height(&self, txid: TxId) -> Result<Option<BlockHeight>, Self::Error>;
+
+    fn get_address<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        account: AccountId,
+    ) -> Result<Option<PaymentAddress>, Self::Error>;
+
+    fn get_extended_full_viewing_keys<P: consensus::Parameters>(
+        &self,
+        params: &P,
+    ) -> Result<Vec<ExtendedFullViewingKey>, Self::Error>;
+
+    fn is_valid_account_extfvk<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        account: AccountId,
+        extfvk: &ExtendedFullViewingKey,
+    ) -> Result<bool, Self::Error>;
+
+    fn get_balance(&self, account: AccountId) -> Result<Amount, Self::Error>;
+
+    fn get_verified_balance(
+        &self,
+        account: AccountId,
+        anchor_height: BlockHeight,
+    ) -> Result<Amount, Self::Error>;
+
+    fn get_received_memo_as_utf8(
+        &self,
+        id_note: Self::NoteRef,
+    ) -> Result<Option<String>, Self::Error>;
+
+    fn get_sent_memo_as_utf8(&self, id_note: Self::NoteRef) -> Result<Option<String>, Self::Error>;
+
+    fn get_commitment_tree(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Option<CommitmentTree<Node>>, Self::Error>;
+
+    fn get_witnesses(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Vec<(Self::NoteRef, IncrementalWitness<Node>)>, Self::Error>;
+
+    fn get_nullifiers(&self) -> Result<Vec<(Vec<u8>, AccountId)>, Self::Error>;
+
+    fn select_spendable_notes(
+        &self,
+        account: AccountId,
+        target_value: Amount,
+        anchor_height: BlockHeight,
+    ) -> Result<Vec<SpendableNote>, Self::Error>;
+
+    fn get_update_ops(&self) -> Result<Self::UpdateOps, Self::Error>;
```

Comment:
After some consideration, I think this shouldn't be a part of the WalletRead interface, but I'm not sure where it's best to put it. I think the better approach is going to be to add a `WalletRead` implementation for the write handle, and then I will constrain locations that need both read and write to both traits.

Reply:
Moved out of the interface.